torchgeo/datasets/quakeset.py

```diff
@@ -0,0 +1,291 @@
+
+        Raises:
+            AssertionError: If ``split`` argument is invalid.
+            DatasetNotFoundError: If dataset is not found and *download* is False.
+        """
+        assert split in self.splits
+
+        self.root = root
+        self.split = split
+        self.transforms = transforms
+        self.download = download
+        self.checksum = checksum
+        self.filepath = os.path.join(root, self.filename)
+
+        self._verify()
+
+        try:
+            import h5py  # noqa: F401
+        except ImportError:
+            raise ImportError(
+                "h5py is not installed and is required to use this dataset"
+            )
+
+        self.data = self._load_data()
+
+    def __getitem__(self, index: int) -> dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            sample containing image and mask
+        """
+        image = self._load_image(index)
+        label = torch.tensor(self.data[index]["label"])
+        magnitude = torch.tensor(self.data[index]["magnitude"])
+
+        sample = {"image": image, "label": label, "magnitude": magnitude}
+
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        """Return the number of data points in the dataset.
+
+        Returns:
+            length of the dataset
+        """
+        return len(self.data)
+
+    def _load_data(self) -> list[dict[str, Any]]:
+        """Return the metadata for a given split.
+
+        Returns:
+            the sample keys, patches, images, labels, and magnitudes
+        """
+        import h5py
+
+        f = h5py.File(self.filepath)
+
+        data = []
+        for k in sorted(f.keys()):
+            if f[k].attrs["split"] != self.splits[self.split]:
+                continue
+
+            for patch in sorted(f[k].keys()):
+                if patch not in ["x", "y"]:
+                    # positive sample
+                    magnitude = float(f[k].attrs["magnitude"])
+                    data.append(
+                        dict(
+                            key=k,
+                            patch=patch,
+                            images=("pre", "post"),
+                            label=1,
+                            magnitude=magnitude,
+                        )
+                    )
+
+                    # hard negative sample
+                    if "before" in f[k][patch].keys():
+                        data.append(
+                            dict(
+                                key=k,
+                                patch=patch,
+                                images=("before", "pre"),
+                                label=0,
+                                magnitude=0.0,
+                            )
+                        )
+        f.close()
```

Comment:
Can we use `with h5py.File(...) as f:`?

Reply:
Seems like that just adds more indentation

Reply:
It also helps prevent file handle leaks if the process dies inside the loop.

Reply:
Indentation can be reduced by inverting the if-statement on line 152, or by saving the dict to a variable before appending.